hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/server/TopicMetadataLoadingStartupHook.java

```diff
@@ -0,0 +1,208 @@
+                config.getIntProperty(FRONTEND_STARTUP_TOPIC_METADATA_LOADING_THREAD_POOL_SIZE));
+    }
+
+    public TopicMetadataLoadingStartupHook(BrokerMessageProducer brokerMessageProducer,
+                                           TopicsCache topicsCache,
+                                           GroupRepository groupRepository,
+                                           TopicRepository topicRepository,
+                                           int retryCount,
+                                           long retryInterval,
+                                           int threadPoolSize) {
+        this.brokerMessageProducer = brokerMessageProducer;
+        this.topicsCache = topicsCache;
+        this.groupRepository = groupRepository;
+        this.topicRepository = topicRepository;
+        this.retryCount = retryCount;
+        this.retryInterval = retryInterval;
+        this.threadPoolSize = threadPoolSize;
+    }
+
+    @Override
+    public void accept(ServiceLocator serviceLocator) {
+        long start = System.currentTimeMillis();
+        logger.info("Loading topics metadata");
+        ExecutorService executorService = Executors.newFixedThreadPool(threadPoolSize);
+        try {
+            loadTopicsMetadataInExecutor(executorService);
+        } finally {
+            executorService.shutdown();
+        }
+        logger.info("Loading topics metadata done in {}ms", System.currentTimeMillis() - start);
+    }
+
+    private void loadTopicsMetadataInExecutor(ExecutorService executorService) {
+        List<TopicName> topics = getTopics();
+
+        List<LoadingResult> successes = new ArrayList<>();
+        List<LoadingResult> failures = Collections.emptyList();
+
+        int retriesLeft = retryCount + 1;
+        while (topics.size() > 0 && retriesLeft > 0) {
+            retriesLeft--;
+            if (retriesLeft < retryCount) {
+                waitBeforeNextRetry();
+            }
+            List<CompletableFuture<LoadingResult>> futures = submitTopicMetadataLoadingTasks(executorService, topics);
+            List<LoadingResult> allResults = whenAllComplete(futures).join();
+
+            Map<LoadingResult.Result, List<LoadingResult>> groupedResults = getGroupedResults(allResults);
+
+            Optional.ofNullable(groupedResults.get(LoadingResult.Result.SUCCESS)).ifPresent(successes::addAll);
+            failures = Optional.ofNullable(groupedResults.get(LoadingResult.Result.FAILURE)).orElse(Collections.emptyList());
+            topics = failures.stream().map(LoadingResult::getTopicName).collect(toList());
+        }
+        logResultInfo(successes, failures);
+    }
+
+    private List<TopicName> getTopics() {
+        return groupRepository.listGroupNames().stream()
+                .flatMap(group -> topicRepository.listTopicNames(group).stream().map(topic -> new TopicName(group, topic)))
+                .collect(toList());
+    }
+
+    private void waitBeforeNextRetry() {
```

Comment:
hmmm how about using the lib we have in hermes-client instead of writing on retrier? take a look at how it is done